### ISSUE TITLE
Django 1.10 compatibility issue

### ISSUE THREAD

**Issue by kierun** [romgar/django-dirtyfields#78]
I have two models, `Assembly` and `Location`.  An assembly may contain other assemblies so when a location changes, I want to make sure the location of all the sub-assemblies changes as well. The code to do this is:

```
    def save(self, *args, **kwargs):
        """Additions to the default abstract save method."""
        try:
            import ipdb; ipdb.set_trace()  # Used to get debugging info.
            if self.is_dirty(check_relationship=True):
                dirties = self.get_dirty_fields(check_relationship=True)
                if 'locality' in dirties.keys():
                    self.ripple_downwards('locality', self.locality)
        finally:
            super(Assembly, self).save(*args, **kwargs)
```

This used to work fine with `dirtyfields` version 0.8.2. However, with 1.1 it does not. Here is some debugging information:

```
ipdb> self.pk
1
ipdb> sut = Assembly.objects.get(pk=1)
ipdb> sut.pk == self.pk
True
ipdb> self.locality
<Location: Afghanistan>
ipdb> sut.locality is None
True
ipdb> self.get_dirty_fields(check_relationship=True)
{}
```

Am I doing something wrong?

**Comment by kierun** [comment]
Looking into it a little more, it appears to be a problem with Django `1.10.0`.  If I use `1.9.8`, all is fine. Weird.

**Comment by romgar** [comment]
Hi @kierun,

`django-dirty-fields` is not yet compatible with Django `1.10`, that's why you have this issue.
I will work on that quite soon!!

**Comment by kierun** [comment]
@romgar Well, that would indeed explain it. :) Thank you.

If you need someone to test something, please let me know.

**Comment by romgar** [comment]
@kierun you can give a try to `compatibility_with_django_1_10` branch, it should be better!

**Comment by kierun** [comment]
Apologies, did not see this till today. I tried it, HEAD is `commit 707527afda9563cab1c0bc61600ba6efcdd86adb` and I am still getting the same error as above.

**Comment by romgar** [comment]
Can you please create a MR and write a test that reproduces it?

**Comment by kierun** [comment]
For some reason ("Unfortunately, we don’t support that file type."… But it's a ZIP file?), I cannot attach the code. So, here it is in all its gorey details:

```python
import pytest
from django.db import models
from dirtyfields import DirtyFieldsMixin


class Location(DirtyFieldsMixin, models.Model):
    name = models.CharField(max_length=128)


class Assembly(DirtyFieldsMixin, models.Model):
    contained_in = models.ForeignKey('self', null=True, blank=True)
    locality = models.ForeignKey(Location, null=True, blank=True)

    def save(self, *args, **kwargs):
        """Additions to the default abstract save method."""
        try:
            # import ipdb; ipdb.set_trace()
            if self.is_dirty(check_relationship=True):
                dirties = self.get_dirty_fields(check_relationship=True)
                if 'locality' in dirties.keys():
                    self.ripple_downwards('locality', self.locality)
        finally:
            super(Assembly, self).save(*args, **kwargs)

    def ripple_downwards(self, key, value):
        """Sets key to value in all sub assemblies in a recursive way.

        :param String key: A model field.
        :param object value: An object those model's field can be set to.

        """
        for item in self.assembly_set.all():
            setattr(item, key, value)
            item.save()
            item.ripple_downwards(key, value)


@pytest.mark.django_db
def test_ripple_down():
    sut = Assembly.objects.create()
    sub = Assembly.objects.create()
    sub.contained_in = sut

    local = Location.objects.create(name='Afghanistan')
    sut.locality = local
    sut.save()  # This should cause the Location to be passed downwards.
    sut.refresh_from_db()

    sub.refresh_from_db()
    assert local == sub.locality
```

**Comment by kierun** [comment]
As a side note, I am not that experienced with Django so my assumption that the above code is actually how you should do that might be erroneous…

**Comment by romgar** [comment]
I will have a deeper look soon @kierun.
What is sure is that there were some compatibility issues with Django `1.10` anyway :-)

**Comment by kierun** [comment]
@romgar Thank you very much. Let me know if you need anything more from me as I am ready to help.

**Comment by romgar** [comment]
Hi @kierun,

After investigations, I reproduced all issues you have/had:
- Commit `707527afda9563cab1c0bc61600ba6efcdd86adb` is fixing almost all of them,
- From the code you copy/pasted, you forgot a `save()`:

```

@pytest.mark.django_db
def test_ripple_down():
    sut = Assembly.objects.create()
    sub = Assembly.objects.create()
    sub.contained_in = sut

    sub.save()  # HERE: you should save sub object, or it won't be linked with sut

    local = Location.objects.create(name='Afghanistan')
    sut.locality = local
    sut.save()
    sut.refresh_from_db()

    sub.refresh_from_db()
    assert local == sub.locality  # and if you missed the sut <-> sub link, sub can't be updated properly
```

You also don't need to call `is_dirty` and `get_dirty_fields` in your example:
```
def save(self, *args, **kwargs):
    """Additions to the default abstract save method."""
    try:
        # import ipdb; ipdb.set_trace()
        dirties = self.get_dirty_fields(check_relationship=True)
        if 'locality' in dirties.keys():
            self.ripple_downwards('locality', self.locality)
    finally:
        super(Assembly, self).save(*args, **kwargs)
```
Tell me if it sounds better with these feedbacks.

**Comment by kierun** [comment]
Thank you for the code clean up! The `save()` is now much more pythonic.

Yes, I was missing the `sub.save()`!  In my real project, it was there but…

On the real project, somehow `python setup.py install` mixed with the old pip installed version. resulting in  some subtly broken mix of old and new version of `dirtyfield`.  Having sorted that, the tests now pass again.

So, I guess that fixes it.

Many thanks indeed.

**Comment by romgar** [comment]
Version `1.2` released and compatible with Django 1.10, you can now use this one.
